<!DOCTYPE html>
<html lang="en">
<head>
  
    <title>場当たり的な例外処理 :: ご乱心.jp</title>
  
  <meta http-equiv="content-type" content="text/html; charset=utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">
<meta name="description" content="定期的に動かすので、ある程度きちんとプログラムを書かないといけないけど、そんなに丁寧に例外処理を書いている余裕はないぞというケースがある。 そういう時、実行スクリプトでは、このように雑に例外処理を行うこ" />
<meta name="keywords" content="プログラミング、ゲーム、ご乱心" />
<meta name="robots" content="noodp" />
<link rel="canonical" href="/posts/exception/" />




<link rel="stylesheet" href="/assets/style.css">

  <link rel="stylesheet" href="/assets/green.css">



<link rel="stylesheet" href="/style.css">


<link rel="apple-touch-icon-precomposed" sizes="144x144" href="/img/apple-touch-icon-144-precomposed.png">

  <link rel="shortcut icon" href="/img/favicon/green.png">



<meta name="twitter:card" content="summary" />

  <meta name="twitter:site" content="" />

<meta name="twitter:creator" content="nowaai" />


<meta property="og:locale" content="en" />
<meta property="og:type" content="article" />
<meta property="og:title" content="場当たり的な例外処理 :: ご乱心.jp">
<meta property="og:description" content="定期的に動かすので、ある程度きちんとプログラムを書かないといけないけど、そんなに丁寧に例外処理を書いている余裕はないぞというケースがある。 そういう時、実行スクリプトでは、このように雑に例外処理を行うこ" />
<meta property="og:url" content="/posts/exception/" />
<meta property="og:site_name" content="場当たり的な例外処理" />

  
    <meta property="og:image" content="/img/favicon/green.png">
  

<meta property="og:image:width" content="2048">
<meta property="og:image:height" content="1024">


  <meta property="article:published_time" content="2020-02-21 22:37:52 &#43;0900 JST" />












<script async src="https://www.googletagmanager.com/gtag/js?id=UA-158846630-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-158846630-2');
</script>



</head>
<body class="">


<div class="container center headings--one-size">

  <header class="header">
  <div class="header__inner">
    <div class="header__logo">
      <a href="/">
  <div class="logo">
    ご乱心.jp
  </div>
</a>

    </div>
    <div class="menu-trigger">menu</div>
  </div>
  
    <nav class="menu">
  <ul class="menu__inner menu__inner--desktop">
    
      
        
          <li><a href="/about">About</a></li>
        
      
      
    

    
  </ul>

  <ul class="menu__inner menu__inner--mobile">
    
      
        <li><a href="/about">About</a></li>
      
    
    
  </ul>
</nav>

  
</header>


  <div class="content">
    
<div class="post">
  <h1 class="post-title">
    <a href="/posts/exception/">場当たり的な例外処理</a></h1>
  <div class="post-meta">
      
    <span class="post-date">
      2020-02-21
    </span>
    
    
    <span class="post-author">::
      nowaai
    </span>
    
  </div>

  
  <span class="post-tags">
    
    #<a href="/tags/python/">Python</a>&nbsp;
    
  </span>
  

  

  <div class="post-content"><div>
        <p>定期的に動かすので、ある程度きちんとプログラムを書かないといけないけど、そんなに丁寧に例外処理を書いている余裕はないぞというケースがある。</p>
<p>そういう時、実行スクリプトでは、このように雑に例外処理を行うことが多い。
こうしておけば、とりあえずどんな例外もログに吐き出すことができ、未知の例外に遭遇した時はエラーメッセージを見ながら、適宜コードを修正することができる。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#f92672">import</span> traceback


<span style="color:#66d9ef">try</span>:
    a()
    b()
    c()
    d()
    e()
<span style="color:#66d9ef">except</span>:
    msg <span style="color:#f92672">=</span> traceback<span style="color:#f92672">.</span>format_exc()
    mail_logger<span style="color:#f92672">.</span>error(msg)
</code></pre></div><p>ところでDB周りの業務をやっていると、SQLAlchemyを使って、下のように<code>session_scope</code>関数を定義し、色んなところで使いまわす。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#f92672">from</span> contextlib <span style="color:#f92672">import</span> contextmanager


<span style="color:#a6e22e">@contextmanager</span>
<span style="color:#66d9ef">def</span> <span style="color:#a6e22e">session_scope</span>():
    session <span style="color:#f92672">=</span> Session()
    <span style="color:#66d9ef">try</span>:
        <span style="color:#66d9ef">yield</span> session
        session<span style="color:#f92672">.</span>commit()
    <span style="color:#66d9ef">except</span>:
        session<span style="color:#f92672">.</span>rollback()
    <span style="color:#66d9ef">finally</span>:
        session<span style="color:#f92672">.</span>close()

<span style="color:#75715e"># session_scopeはこのように使う</span>
<span style="color:#66d9ef">def</span> <span style="color:#a6e22e">insert</span>(something):
    <span style="color:#66d9ef">with</span> session_scope() <span style="color:#66d9ef">as</span> session:
        session<span style="color:#f92672">.</span>add(something)
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#75715e"># メインの処理はこのような感じ</span>
<span style="color:#66d9ef">try</span>:
    something <span style="color:#f92672">=</span> Something()
    insert(something)
<span style="color:#66d9ef">except</span>:
    msg <span style="color:#f92672">=</span> traceback<span style="color:#f92672">.</span>format_exc()
    mail_logger<span style="color:#f92672">.</span>error(msg)
</code></pre></div><p>しかし、この<code>session_scope</code>を上記のようなtry句内で使うと、DB操作で例外が発生した時に、<code>session_scope</code>内の<code>except</code>、<code>finnaly</code>でプログラムが終了し、肝心の最上位の<code>except</code>句に処理が到達しない。
こうなると、ログが吐き出されないので、困ったことになる。</p>
<p>こういうときは、<code>session_scope</code>内の<code>except</code>句の中で例外を再スローしてやるとよい。
そうすると、きちんと処理が最上位の<code>except</code>にまで到達し、DB操作時に例外が発生した時の状況がログに書き込まれる。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#f92672">from</span> exceptions <span style="color:#f92672">import</span> DBHandleException


<span style="color:#a6e22e">@contextmanager</span>
<span style="color:#66d9ef">def</span> <span style="color:#a6e22e">session_scope</span>():
    session <span style="color:#f92672">=</span> Session()
    <span style="color:#66d9ef">try</span>:
        <span style="color:#66d9ef">yield</span> session
        session<span style="color:#f92672">.</span>commit()
    <span style="color:#66d9ef">except</span>:
        session<span style="color:#f92672">.</span>rollback()
        <span style="color:#75715e"># 例外を再スローする</span>
        <span style="color:#66d9ef">raise</span> DBHandleException
    <span style="color:#66d9ef">finally</span>:
        session<span style="color:#f92672">.</span>close()
</code></pre></div>
      </div></div>
  
  <div class="pagination">
    <div class="pagination__title">
      <span
        class="pagination__title-h">Read other posts</span>
      <hr />
    </div>
    <div class="pagination__buttons">
      
      <span class="button previous">
        <a href="/posts/shadowtactics/">
          <span class="button__icon">←</span>
          <span class="button__text">Shadow Tactics 雑感</span>
        </a>
      </span>
      
      
      <span class="button next">
        <a href="/posts/control/">
          <span class="button__text">Control雑感</span>
          <span class="button__icon">→</span>
        </a>
      </span>
      
    </div>
  </div>
  

  

</div>

  </div>

  
    <footer class="footer">
  <div class="footer__inner">
    
      <div class="copyright copyright--user">
        <span>@nowaai</span>
    
        <span>:: Theme made by <a href="https://twitter.com/panr">panr</a></span>
      </div>
  </div>
</footer>

<script src="/assets/main.js"></script>
<script src="/assets/prism.js"></script>






  
</div>

</body>
</html>
